File: BOJ/Gold IV/BOJ1717.py
#BOJ1717 집합의 표현 20210324 유니온 파인드
import sys
input = sys.stdin.readline

parent = []

def find(x):
    if x == parent[x]:
        return x
    else:
        y = find(parent[x])
        parent[x] = y
        return y

def union(x,y):
    x = find(x)
    y = find(y)
    parent[y] = x

def main():
    global parent
    n,m = map(int,input().rstrip().split())
    parent = [i for i in range(n+1)]
    for ___ in range(m):
        cmd, a, b, = map(int, input().rstrip().split())
        if cmd == 1:
            if find(a) == find(b):
                print("YES")
            else:
                print("NO")
        else:
            union(a,b)
if __name__ == '__main__':
    main()

File: BOJ/Gold V/BOJ1717.java
//BOJ1717 집합의 표현 20210712
import java.io.*;
import java.util.*;

import static java.lang.Math.*;

public class Main {
    private static final BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    private static final BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

    private static int n,m;
    private static int[] parent;

    public static void main(String[] args) throws IOException {
        StringTokenizer stk = new StringTokenizer(br.readLine());
        n = Integer.parseInt(stk.nextToken());
        m = Integer.parseInt(stk.nextToken());

        parent = new int[n+1];
        for(int i = 0; i <= n; i++){
            parent[i] = i;
        }

        for(int i = 0; i < m; i++){
            stk = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(stk.nextToken());
            int b = Integer.parseInt(stk.nextToken());
            int c = Integer.parseInt(stk.nextToken());
            if(a == 0){
                union(b, c);
            }else{
                bw.write(((find(b) == find(c))? "YES" : "NO") + "\n");
            }
        }
        bw.flush();
        bw.close();
    }

    public static int find(int v){
        if(parent[v] == v){
            return v;
        }
        parent[v] = find(parent[v]);
        return parent[v];
    }

    public static void union(int a, int b){
        int ap = find(a);
        int bp = find(b);
        parent[ap] = bp;
    }
}